### DISCUSSION TITLE
The compiler is unable to type-check `Button` this expression in reasonable time;

### DISCUSSION THREAD

**Discussion by saroar** [pointfreeco/swift-composable-architecture#1641] [Q&A]
I am not sure it bug or my code issue 
issue is when I am trying send action from view inside  multiple bool condition
WordTodoView.swift line number 192

<img width="952" alt="Screenshot 2022-11-09 at 17 03 06" src="https://user-images.githubusercontent.com/8770772/200851422-0b0a3506-19f6-4a20-9016-e6eea302f729.png">

**Comment by tgrapperon** [comment]
Hey @saroar! The `ViewBuilder` can sometimes have difficulties resolving complex views, especially when multiple conditional are chained. In this case, you can probably help him a little by wrapping your `if viewStore…` block in a `Group`.
```swift
Group {
  if viewStore … {
    …
  }
}
```

**Comment by mbrandonw** [comment]
Another option, in addition to @tgrapperon's suggestion, is to use `@ObservedObject` instead of `WithViewStore`, which we describe [here](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance#Compiler-performance).

I also tried compiling this code pointed to a [branch](https://github.com/pointfreeco/swift-composable-architecture/compare/remove-withviewstore-deprecations) with the 12 deprecated inits on `WithViewStore` removed and somehow it didn't make a difference. Has me second guessing whether those inits were ever problematic. I thought we had confirmed that in a sample project.

**Reply by tgrapperon** [reply]
@mbrandonw I don't thing there ever was a direct proof. According to my tests at the time, I think that locking `WithViewStore.Content` to a `View` greatly improves the situation. This is not possible as long as the other overloads are still present of course.

**Reply by mbrandonw** [reply]
Ah, maybe that's what I'm thinking of. We did have someone share a private repo with us many months ago so that we could look into the problem, and pointing the project to a particular branch did fix the problem. Let me try that real quick.

**Reply by mbrandonw** [reply]
Wow, bringing back `Content: View` _and_ removing the deprecated non-`observe:` inits still doesn't fix the problem. [Here's](https://github.com/pointfreeco/swift-composable-architecture/compare/remove-withviewstore-deprecations) the branch I'm targeting.

**Reply by tgrapperon** [reply]
This is also really odd: on `main`, if you remove the reference to `viewStore` in the `Button`'s action, it builds:
```swift
let takeQuiz: () -> Void = { viewStore.send(.takeQuiz) }
if viewStore.isActive && !viewStore.isComplete {
  HStack {
    Button {
      takeQuiz()
    } label: {
      Text("Take Quiz!")
    }
  }
}
```
If you simply replace the `Button`'s action by a lone `viewStore` that does nothing, it doesn't build anymore. I'm not sure what's going on here, because this thing doesn't change the generic signature of the `View`.

**Reply by tgrapperon** [reply]
This is even weirder. This builds:
```swift
if true {
  let _ = viewStore
  HStack {
    Color.red
  }
}       
```
This also builds, but it takes considerably longer:
```swift
if true {
  HStack {
    let _ = viewStore
    Color.red
  }
}  
```
If we replace `let _ = viewStore` by `let action: () -> Void = { viewStore.send(.takeQuiz) }`, it still builds in the first case, but fails in the second. If we replace by `let action: () -> Void = { }`, it builds instantly in both cases.

**Reply by mbrandonw** [reply]
While messing around with this we discovered that giving an explicit type to the view store vended by `WithViewStore` also greatly helps:

```swift
WithViewStore(self.store, observe: { $0 }) { (viewStore: ViewStoreOf<WordTodo>) in
  …
}
```

So that is another alternative if one does not want to use `@ObservedObject`. We are going to update the performance article.

**Reply by Alex293** [reply]
I tend to rely on the explicit type when things get out of control and it’s working great

**Comment by saroar** [comment]
I am big fan of  `@ObservedObject`  yes it fixe issue @mbrandonw @tgrapperon @stephencelis thanks

**Reply by Alex293** [reply]
Be careful though as this will redraw the entire view IIRC

**Reply by saroar** [reply]
what is `view IIRC` ? @Alex293 
so if I will use inside view `ViewState` ?

**Reply by Alex293** [reply]
If I remember correctly, sorry for that. Using @Viewstate at the view level will redraw the entier view. You need to make a subview if you want to avoid that. @mbrandonw

**Reply by tgrapperon** [reply]
In any case, you should always use a `ViewState` if your state contains superfluous data for the `View`. If the `ViewStore` is installed as an `@ObservedObject`, it will indeed invalidate the whole `body`, but this is very similar to what already happens if your `WithViewStore` is a top-level `View` in `body`, so it isn't really worse in terms of performance (it is even very slightly better than a top-level `WithViewStore` given the current implementation).
What this prevents is narrowing the `ViewStore`'s reach to only a portion of your `View`, which is something possible with a `WithViewStore` used in a non top-level position. But this is less frequently used.
